### DISCUSSION TITLE
Windows linking issues

### DISCUSSION THREAD

**Discussion by sbonar** [libuv/libuv#4064] [Q&A] [answered]
Hopefully I'm not over staying my welcome with all my questions. :-)

I'm getting a lot of Linker warnings and errors when trying to statically link libuv.lib to my application.

> 24>LINK : warning LNK4217: symbol '_errno' defined in 'libucrt.lib(errno.obj)' is imported by 'libuv.lib(uv-common.obj)' in function 'uv__free'
> 24>LINK : warning LNK4217: symbol '_errno' defined in 'libucrt.lib(errno.obj)' is imported by 'libuv.lib(thread.obj)' in function 'uv_sem_wait'
> 24>LINK : warning LNK4286: symbol '_errno' defined in 'libucrt.lib(errno.obj)' is imported by 'libuv.lib(pipe.obj)'
> 24>LINK : warning LNK4286: symbol '_errno' defined in 'libucrt.lib(errno.obj)' is imported by 'libuv.lib(fs.obj)'
> 24>LINK : warning LNK4217: symbol '__stdio_common_vfprintf' defined in 'libucrt.lib(output.obj)' is imported by 'libuv.lib(uv-common.obj)' in function '_vfprintf_l'
> 24>LINK : warning LNK4286: symbol '__stdio_common_vfprintf' defined in 'libucrt.lib(output.obj)' is imported by 'libuv.lib(error.obj)'
> 24>LINK : warning LNK4217: symbol '__stdio_common_vfprintf' defined in 'libucrt.lib(output.obj)' is imported by 'libuv.lib(pipe.obj)' in function 'uv_pipe'
> 24>LINK : warning LNK4217: symbol '__stdio_common_vsprintf' defined in 'libucrt.lib(output.obj)' is imported by 'libuv.lib(uv-common.obj)' in function 'vsnprintf'
> 24>LINK : warning LNK4286: symbol '__stdio_common_vsprintf' defined in 'libucrt.lib(output.obj)' is imported by 'libuv.lib(inet.obj)'
> 24>LINK : warning LNK4217: symbol '__stdio_common_vsprintf' defined in 'libucrt.lib(output.obj)' is imported by 'libuv.lib(util.obj)' in function 'uv_getrusage'
> 24>LINK : warning LNK4217: symbol '__stdio_common_vsprintf' defined in 'libucrt.lib(output.obj)' is imported by 'libuv.lib(pipe.obj)' in function 'close_pipe'
> 24>LINK : warning LNK4217: symbol 'atoi' defined in 'libucrt.lib(atox.obj)' is imported by 'libuv.lib(uv-common.obj)' in function 'uv_ip6_addr'
> 24>LINK : warning LNK4286: symbol 'atoi' defined in 'libucrt.lib(atox.obj)' is imported by 'libuv.lib(threadpool.obj)'
> 24>LINK : warning LNK4217: symbol '_set_invalid_parameter_handler' defined in 'libucrt.lib(invalid_parameter.obj)' is imported by 'libuv.lib(core.obj)' in function 'uv__init'
> 24>LINK : warning LNK4286: symbol '_get_osfhandle' defined in 'libucrt.lib(osfinfo.obj)' is imported by 'libuv.lib(fs.obj)'
> 24>LINK : warning LNK4217: symbol '_get_osfhandle' defined in 'libucrt.lib(osfinfo.obj)' is imported by 'libuv.lib(process-stdio.obj)' in function 'uv__get_osfhandle'
> 24>LINK : warning LNK4217: symbol '_get_osfhandle' defined in 'libucrt.lib(osfinfo.obj)' is imported by 'libuv.lib(handle.obj)' in function 'uv__get_osfhandle'
> 24>LINK : warning LNK4286: symbol '_get_osfhandle' defined in 'libucrt.lib(osfinfo.obj)' is imported by 'libuv.lib(pipe.obj)'
> 24>LINK : warning LNK4286: symbol '_get_osfhandle' defined in 'libucrt.lib(osfinfo.obj)' is imported by 'libuv.lib(tty.obj)'
> 24>LINK : warning LNK4286: symbol '_get_osfhandle' defined in 'libucrt.lib(osfinfo.obj)' is imported by 'libuv.lib(poll.obj)'
> 24>LINK : warning LNK4217: symbol '_open_osfhandle' defined in 'libucrt.lib(osfinfo.obj)' is imported by 'libuv.lib(handle.obj)' in function 'uv_open_osfhandle'
> 24>LINK : warning LNK4286: symbol '_open_osfhandle' defined in 'libucrt.lib(osfinfo.obj)' is imported by 'libuv.lib(pipe.obj)'
> 24>LINK : warning LNK4286: symbol '_open_osfhandle' defined in 'libucrt.lib(osfinfo.obj)' is imported by 'libuv.lib(fs.obj)'
> 24>LINK : warning LNK4217: symbol 'abort' defined in 'libucrt.lib(abort.obj)' is imported by 'libuv.lib(fs-poll.obj)' in function 'uv_fs_poll_start'
> 24>LINK : warning LNK4286: symbol 'abort' defined in 'libucrt.lib(abort.obj)' is imported by 'libuv.lib(winsock.obj)'
> 24>LINK : warning LNK4217: symbol 'abort' defined in 'libucrt.lib(abort.obj)' is imported by 'libuv.lib(signal.obj)' in function 'uv__signals_init'
> 24>LINK : warning LNK4286: symbol 'abort' defined in 'libucrt.lib(abort.obj)' is imported by 'libuv.lib(threadpool.obj)'
> 24>LINK : warning LNK4286: symbol 'abort' defined in 'libucrt.lib(abort.obj)' is imported by 'libuv.lib(error.obj)'
> 24>LINK : warning LNK4217: symbol 'abort' defined in 'libucrt.lib(abort.obj)' is imported by 'libuv.lib(tty.obj)' in function 'uv__tty_read_start'
> 24>LINK : warning LNK4217: symbol 'abort' defined in 'libucrt.lib(abort.obj)' is imported by 'libuv.lib(handle.obj)' in function 'uv_close'
> 24>LINK : warning LNK4217: symbol 'abort' defined in 'libucrt.lib(abort.obj)' is imported by 'libuv.lib(udp.obj)' in function 'uv_udp_set_multicast_interface'
> 24>LINK : warning LNK4217: symbol 'abort' defined in 'libucrt.lib(abort.obj)' is imported by 'libuv.lib(thread.obj)' in function 'uv_rwlock_init'
> 24>LINK : warning LNK4217: symbol 'abort' defined in 'libucrt.lib(abort.obj)' is imported by 'libuv.lib(tcp.obj)' in function 'uv_socketpair'
> 24>LINK : warning LNK4217: symbol '__stdio_common_vswprintf' defined in 'libucrt.lib(output.obj)' is imported by 'libuv.lib(util.obj)' in function '_vsnwprintf_l'
> 24>LINK : warning LNK4286: symbol '__stdio_common_vswprintf' defined in 'libucrt.lib(output.obj)' is imported by 'libuv.lib(fs.obj)'
> 24>LINK : warning LNK4217: symbol '__stdio_common_vswprintf' defined in 'libucrt.lib(output.obj)' is imported by 'libuv.lib(fs-event.obj)' in function 'uv__relative_path'
> 24>LINK : warning LNK4217: symbol 'wcslen' defined in 'libucrt.lib(strnlen.obj)' is imported by 'libuv.lib(fs-event.obj)' in function 'uv_fs_event_start'
> 24>LINK : warning LNK4217: symbol 'wcslen' defined in 'libucrt.lib(strnlen.obj)' is imported by 'libuv.lib(util.obj)' in function 'uv_os_environ'
> 24>LINK : warning LNK4217: symbol 'wcslen' defined in 'libucrt.lib(strnlen.obj)' is imported by 'libuv.lib(pipe.obj)' in function 'uv__create_pipe_pair'
> 24>LINK : warning LNK4286: symbol 'wcslen' defined in 'libucrt.lib(strnlen.obj)' is imported by 'libuv.lib(process.obj)'
> 24>LINK : warning LNK4217: symbol 'wcslen' defined in 'libucrt.lib(strnlen.obj)' is imported by 'libuv.lib(fs.obj)' in function 'fs__rmdir'
> 24>LINK : warning LNK4217: symbol 'wcsncmp' defined in 'libucrt.lib(wcsncmp.obj)' is imported by 'libuv.lib(util.obj)' in function 'uv_os_uname'
> 24>LINK : warning LNK4217: symbol 'wcsncmp' defined in 'libucrt.lib(wcsncmp.obj)' is imported by 'libuv.lib(fs.obj)' in function 'fs__chmod'
> 24>LINK : warning LNK4217: symbol '_beginthreadex' defined in 'libucrt.lib(thread.obj)' is imported by 'libuv.lib(thread.obj)' in function 'uv_thread_create_ex'
> 24>LINK : warning LNK4217: symbol 'getenv' defined in 'libucrt.lib(getenv.obj)' is imported by 'libuv.lib(threadpool.obj)' in function 'init_threads'
> 24>LINK : warning LNK4217: symbol '__acrt_iob_func' defined in 'libucrt.lib(_file.obj)' is imported by 'libuv.lib(error.obj)' in function 'uv_fatal_error'
> 24>LINK : warning LNK4286: symbol '__acrt_iob_func' defined in 'libucrt.lib(_file.obj)' is imported by 'libuv.lib(pipe.obj)'
> 24>LINK : warning LNK4217: symbol '_close' defined in 'libucrt.lib(close.obj)' is imported by 'libuv.lib(pipe.obj)' in function 'uv_pipe'
> 24>LINK : warning LNK4286: symbol '_close' defined in 'libucrt.lib(close.obj)' is imported by 'libuv.lib(fs.obj)'
> 24>LINK : warning LNK4286: symbol '_close' defined in 'libucrt.lib(close.obj)' is imported by 'OLDNAMES.lib(close.obi)'
> 24>LINK : warning LNK4217: symbol '__stdio_common_vsnwprintf_s' defined in 'libucrt.lib(output.obj)' is imported by 'libuv.lib(process.obj)' in function '_vsnwprintf_s_l'
> 24>LINK : warning LNK4217: symbol 'qsort' defined in 'libucrt.lib(qsort.obj)' is imported by 'libuv.lib(process.obj)' in function 'make_program_env'
> 24>LINK : warning LNK4217: symbol 'wcspbrk' defined in 'libucrt.lib(wcspbrk.obj)' is imported by 'libuv.lib(process.obj)' in function 'quote_cmd_arg'
> 24>LINK : warning LNK4217: symbol '_wcsnicmp' defined in 'libucrt.lib(wcsnicmp.obj)' is imported by 'libuv.lib(process.obj)' in function 'search_path_join_test'
> 24>LINK : warning LNK4286: symbol '_wcsnicmp' defined in 'libucrt.lib(wcsnicmp.obj)' is imported by 'libuv.lib(fs-event.obj)'
> 24>LINK : warning LNK4217: symbol '__doserrno' defined in 'libucrt.lib(errno.obj)' is imported by 'libuv.lib(fs.obj)' in function 'fs__rmdir'
> 24>LINK : warning LNK4217: symbol 'wcstombs' defined in 'libucrt.lib(wcstombs.obj)' is imported by 'libuv.lib(fs.obj)' in function 'fs__mktemp'
> 24>LINK : warning LNK4217: symbol '_read' defined in 'libucrt.lib(read.obj)' is imported by 'libuv.lib(fs.obj)' in function 'fs__sendfile'
> 24>LINK : warning LNK4217: symbol '_write' defined in 'libucrt.lib(write.obj)' is imported by 'libuv.lib(fs.obj)' in function 'fs__sendfile'
> 24>LINK : warning LNK4217: symbol '_lseeki64' defined in 'libucrt.lib(lseek.obj)' is imported by 'libuv.lib(fs.obj)' in function 'fs__sendfile'
> 24>LINK : warning LNK4286: symbol '_umask' defined in 'libucrt.lib(umask.obj)' is imported by 'OLDNAMES.lib(umask.obi)'
> 24>LINK : warning LNK4286: symbol '_wcsdup' defined in 'libucrt.lib(wcsdup.obj)' is imported by 'OLDNAMES.lib(wcsdup.obi)'
> 24>libuv.lib(fs.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(fs-event.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(fs-poll.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(process-stdio.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(tty.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(poll.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(loop-watcher.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(process.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(signal.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(threadpool.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(async.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(pipe.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(util.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(thread.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(stream.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(tcp.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(uv-common.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(core.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(handle.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(udp.obj) : error LNK2001: unresolved external symbol __imp__wassert
> 24>libuv.lib(core.obj) : error LNK2019: unresolved external symbol __imp__CrtSetReportHook referenced in function uv__init
> 24>libuv.lib(util.obj) : error LNK2019: unresolved external symbol __imp_GetUserProfileDirectoryW referenced in function uv__getpwuid_r
> 24>libuv.lib(pipe.obj) : error LNK2019: unresolved external symbol __imp_close referenced in function close_pipe
> 24>libuv.lib(tty.obj) : error LNK2001: unresolved external symbol __imp_close
> 24>OLDNAMES.lib(close.obi) : error LNK2001: unresolved external symbol __imp_close
> 24>libuv.lib(process.obj) : error LNK2019: unresolved external symbol __imp_wcscpy referenced in function make_program_args
> 24>libuv.lib(process.obj) : error LNK2019: unresolved external symbol __imp_wcsncpy referenced in function search_path_join_test
> 24>libuv.lib(fs.obj) : error LNK2001: unresolved external symbol __imp_wcsncpy
> 24>libuv.lib(fs-event.obj) : error LNK2001: unresolved external symbol __imp_wcsncpy
> 24>libuv.lib(process.obj) : error LNK2019: unresolved external symbol __imp_wcsrev referenced in function quote_cmd_arg
> 24>OLDNAMES.lib(wcsrev.obi) : error LNK2001: unresolved external symbol __imp_wcsrev
> 24>libuv.lib(fs.obj) : error LNK2019: unresolved external symbol __imp__wrmdir referenced in function fs__rmdir
> 24>libuv.lib(fs.obj) : error LNK2019: unresolved external symbol __imp__wchmod referenced in function fs__chmod
> 24>libuv.lib(fs.obj) : error LNK2019: unresolved external symbol __imp_umask referenced in function fs__open
> 24>OLDNAMES.lib(umask.obi) : error LNK2001: unresolved external symbol __imp_umask
> 24>libuv.lib(fs-event.obj) : error LNK2019: unresolved external symbol __imp_wcsdup referenced in function uv__split_path
> 24>OLDNAMES.lib(wcsdup.obi) : error LNK2001: unresolved external symbol __imp_wcsdup
> 24>OLDNAMES.lib(wcsrev.obi) : error LNK2001: unresolved external symbol __imp__wcsrev
> 

What did I do wrong?

**Comment by sbonar** [comment, marked as answer]
Needed to make sure I was compiling libuv with /MT(d) instead of /MD(d)

**Reply by ChadHartman** [reply]
@sbonar How do you compile with `/MT(d)` when using cmake on windows?

**Reply by bnoordhuis** [reply]
@ChadHartman https://cmake.org/cmake/help/latest/prop_tgt/MSVC_RUNTIME_LIBRARY.html
